warehouse/organizations/models.py

```diff
@@ -0,0 +1,274 @@
+#         try:
+#             return (
+#                 self.request.db.query(Organization)
+#                 .filter(
+#                     Organization.normalized_name
+#                     == func.normalize_pep426_name(organization)
+#                 )
+#                 .one()
+#             )
+#         except NoResultFound:
+#             raise KeyError from None
+
+
+# TODO: Determine if this should also utilize SitemapMixin and TwoFactorRequireable
+# class Project(SitemapMixin, TwoFactorRequireable, db.Model):
+class Organization(db.Model):
+    __tablename__ = "organizations"
+    __table_args__ = (
+        CheckConstraint(
+            "name ~* '^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$'::text",
+            name="organizations_valid_name",
+        ),
+    )
+
+    __repr__ = make_repr("name")
+
+    name = Column(Text, nullable=False)
+    normalized_name = orm.column_property(func.normalize_pep426_name(name))
+    display_name = Column(Text, nullable=False)
+    orgtype = Column(
+        Enum(OrganizationType, values_callable=lambda x: [e.value for e in x]),
+        nullable=False,
+    )
+    link_url = Column(URLType, nullable=False)
+    description = Column(Text, nullable=False)
+    is_active = Column(Boolean, nullable=False, default=False)
+    is_approved = Column(Boolean)
+    created = Column(
+        DateTime(timezone=False),
+        nullable=False,
+        server_default=sql.func.now(),
+        index=True,
+    )
+    date_approved = Column(
+        DateTime(timezone=False),
+        nullable=True,
+        onupdate=func.now(),
+    )
+
+    # TODO: Determine if cascade applies to any of these relationships
+    users = orm.relationship(
+        User, secondary=OrganizationRole.__table__, backref="organizations"  # type: ignore # noqa
+    )
+    projects = orm.relationship(
+        "Project", secondary=OrganizationProject.__table__, backref="organizations"  # type: ignore # noqa
+    )
+
+    events = orm.relationship(
+        "OrganizationEvent",
+        backref="organization",
+        cascade="all, delete-orphan",
+        lazy=True,
+    )
+
+    # TODO:
+    #    def __acl__(self):
+
+    def record_event(self, *, tag, ip_address, additional):
+        session = orm.object_session(self)
+        event = OrganizationEvent(
+            organization=self, tag=tag, ip_address=ip_address, additional=additional
+        )
+        session.add(event)
+        session.flush()
+
+        return event
+
+
+class OrganizationEvent(db.Model):
+    __tablename__ = "organization_events"
+
+    organization_id = Column(
+        UUID(as_uuid=True),
+        ForeignKey("organizations.id", deferrable=True, initially="DEFERRED"),
+        nullable=False,
+        index=True,
+    )
+    tag = Column(String, nullable=False)
+    time = Column(DateTime, nullable=False, server_default=sql.func.now())
+    ip_address = Column(String, nullable=False)
+    additional = Column(JSONB, nullable=True)
```

Comment:
I'd like to stave off an explosion of various `*Event` classes by using generic events instead. I'm getting #8324 up to date and would like to merge that first and then refactor this PR to follow that pattern, if that seems reasonable.

Reply:
@di That seems very reasonable, @sterbo had a similar thought about having a generic `Event` class and your PR looks like a clean way of doing it. We’ll use your changes from #8324 once it’s merged.

Reply:
Merged!

Reply:
Thanks!